rfcs/20190729-keras-preprocessing-redesign.md

```diff
@@ -0,0 +1,351 @@
+          Since the pipeline processes batches of images that must all have the same size,
+          this must be provided.
+        shuffle: Whether to shuffle the data (default: True)
+            If set to False, sorts the data in alphanumeric order.
+        seed: Optional random seed for shuffling and transformations.
+        validation_split: Optional float between 0 and 1,
+            fraction of data to reserve for validation.
+        subset: One of "training" or "validation". Only used if `validation_split` is set.
+    """
+
+def preview(self, data, save_to_directory=None, save_prefix=None, save_format='png'):
+    """Enables users to preview the image augmentation configuration.
+
+    # Arguments
+        data: Image data. Could be strings (a list of image paths), a list of PIL image instances,
+            a list of arrays, or a list of eager tensors.
+        save_to_directory: Directory to save transformed images. Mandatory if not in a notebook.
+            If in a notebook and this is not specified, images are displayed in-line.
+        save_prefix: String, filename prefix for saved images.
+        save_format: String, extension for saved images.
+    """
+```
+
+**Note:** `from_arrays` is not included since it is possible to transform Numpy data simply by calling the `ImagePipeline` object (like a layer).
+
+
+### Layers
+
+The new data augmentation layers will inherit `keras.layers.Layer` and work in a similar way.
+
+```python
+Resizing(height, width)  # Resize while distorting aspect ratio
+CenterCrop(height, width)  # Resize without distorting aspect ratio
+Rescaling(value)  # Divide by `value`
+RandomFlip(horizontal=False, vertical=False, seed=None)
+RandomTranslation(amplitude=0., fill_mode='constant', fill_value=0., seed=None)
+RandomRotation(amplitude=0., fill_mode='constant', fill_value=0., seed=None)
+RandomZoom(amplitude=0., fill_mode='constant', fill_value=0., seed=None)
+RandomBrightness(amplitude=0., seed=None)
+RandomContrast(amplitude=0., seed=None)
+RandomSaturation(amplitude=0., seed=None)
+RandomWidth(amplitude=0., seed=None)  # Expand / shrink width while distorting aspect ratio
+RandomHeight(amplitude=0., seed=None)  # Expand / shrink height while distorting aspect ratio
+```
+
+The `amplitude` argument may be:
+- a positive float: it is understood as "fraction of total" (total is the current width, or height, or 180 degrees in the case `RandomRotation`). E.g. `0.2` results in variations in the [-20%, +20%] range. If larger than 1, it is rounded to one for the lower boundary (but not the higher boundary).
+- a tuple of 2 positive floats: understood as a fractional range, e.g. `(0.2, 0.4)` is interpreted as the [-20%, +40%] range. The first float may not be larger than 1.
+
+To do a random center crop that zooms in and discards part of the image, you would do:
+
+```python
+preprocessor = ImagePipeline([
+  RandomZoom([0., 0.2]),
+  CenterCrop(height, width),
+])
+```
```

Comment:
Does this allow injecting custom augmentation? For example, suppose I want to apply Gaussian Blur or channel wise contrast, can I inject that into it directly?

Reply:
Yes you just need to create your own layer.

Reply:
And a single layer can have multiple transformations inside the `call` which can be defined by other libraries like `imguag` or `albumnetation`, right?

Reply:
Yes, that's right. Custom layers offer you great flexibility to implement your own transformations. But note that all transformations should be defined using TF ops if you want performance. Otherwise you'd have to step out of graph execution (it's still technically doable, though).